### ISSUE TITLE
Human readable errors for JSON Schema oneOf handling

### ISSUE THREAD

**Issue by whitlockjc** [apigee-127/sway#15]
There are a number of places in the Swagger JSON Schema file where `oneOf` is used *(parameters for example)* and when the JSON Schema validator produces its error, it is some cryptic error like this: `Data does not match any schemas from 'oneOf'`.  This was also reported in https://github.com/apigee-127/swagger-tools/issues/200  What we need to do is special-case these situations to produce a better, human readable error instead of just providing the error from our JSON Schema validator *(https://github.com/zaggino/z-schema)*

**Comment by whitlockjc** [comment]
The handling of this can be done really simply, changing the error message from `Data does not match any schemas from 'oneOf'` to `Not a valid {definition type} definition`.  This approach would still require you to look into the `inner` property of the error to construct *why* the definition is not valid but at this point, you would have a human readable error and enough information to provide more detail to help the user along.  That is my plan for the first pass for this, we can improve more after we have that in place.

**Comment by mohsen1** [comment]
What is possible is looking at current object error being thrown from and guess what is going wrong. A good example is when a path parameter does not have in: required. It can be done in "semantic validation" phase.

**Comment by whitlockjc** [comment]
First pass is done.  I took the path describe above.  It works but it's not very specific other than telling you that you've tried to define some object and your definition is invalid.

Anytime you have a `oneOf` failure, you have N different schemas that all have their own reason(s) for not matching.  Missing properties, wrong type, enums, etc.  To have some way to just collapse all of those possibilities into some singular human readable error is impossible.  I think the easiest thing would be to have a human readable *"You've tried to describe a {type} but it is not valid"* like the commit for this issue does and figure out a better way to render the `inner` property of these `oneOf` failures so that someone could say *"Well, I tried to define a definition and I see that I need a 'description' property but it is missing."*

I'll keep thinking of course but at least we have something better than what we started with.
